File: v4/rust/iprange-livedb/src/recovery/budget.rs
use std::path::PathBuf;

use crate::error::{Error, Result};

/// Maximum simultaneously retained resources for one recovery operation.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RecoveryBudget {
    pub max_heap_bytes: u64,
    pub max_output_pages: u64,
    pub max_open_files: u32,
    pub max_scratch_bytes: u64,
    pub max_scratch_files: u32,
    pub scratch_directory: Option<PathBuf>,
}

impl RecoveryBudget {
    /// Recovery budget which forbids external scratch files.
    pub const fn heap_only(
        max_heap_bytes: u64,
        max_output_pages: u64,
        max_open_files: u32,
    ) -> Self {
        Self {
            max_heap_bytes,
            max_output_pages,
            max_open_files,
            max_scratch_bytes: 0,
            max_scratch_files: 0,
            scratch_directory: None,
        }
    }

    pub(crate) fn validate(&self) -> Result<()> {
        if self.max_open_files < 2 {
            return Err(Error::BudgetExceeded(
                "recovery requires source and output files",
            ));
        }
        if self.max_output_pages < 2 {
            return Err(Error::BudgetExceeded("recovery output pages"));
        }
        let scratch_limits = self.max_scratch_bytes != 0 && self.max_scratch_files != 0;
        if scratch_limits != self.scratch_directory.is_some()
            || (self.max_scratch_bytes == 0) != (self.max_scratch_files == 0)
        {
            return Err(Error::InvalidArgument(
                "recovery scratch path and limits must be supplied together",
            ));
        }
        Ok(())
    }
}

File: v4/go/internal/recovery/budget.go
package recovery

// Recovery budget (Rust recovery/budget.rs RecoveryBudget): the
// maximum simultaneously retained resources of one recovery
// operation. Scratch limits are accepted and validated together with
// the scratch directory, and a supplied scratch budget enables the
// file-backed page table and the external sort.

import "github.com/firehol/iprange/v4/go/internal/format"

// RecoveryBudget bounds one recovery operation (Rust RecoveryBudget).
type RecoveryBudget struct {
	MaxHeapBytes     uint64
	MaxOutputPages   uint64
	MaxOpenFiles     uint32
	MaxScratchBytes  uint64
	MaxScratchFiles  uint32
	ScratchDirectory string
}

// HeapOnly builds a recovery budget which forbids external scratch
// files (Rust RecoveryBudget::heap_only).
func HeapOnly(maxHeapBytes uint64, maxOutputPages uint64, maxOpenFiles uint32) *RecoveryBudget {
	return &RecoveryBudget{
		MaxHeapBytes:   maxHeapBytes,
		MaxOutputPages: maxOutputPages,
		MaxOpenFiles:   maxOpenFiles,
	}
}

// validate checks the budget invariants (Rust RecoveryBudget::
// validate): source and output files, a minimum output extent, and
// scratch limits supplied together with the scratch directory.
func (b *RecoveryBudget) validate() error {
	if b.MaxOpenFiles < 2 {
		return &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery requires source and output files"}
	}
	if b.MaxOutputPages < 2 {
		return &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery output pages"}
	}
	scratchLimits := b.MaxScratchBytes != 0 && b.MaxScratchFiles != 0
	if scratchLimits != (b.ScratchDirectory != "") || (b.MaxScratchBytes == 0) != (b.MaxScratchFiles == 0) {
		return &format.Error{Code: format.CodeInvalidArgument, Detail: "recovery scratch path and limits must be supplied together"}
	}
	return nil
}
